File: control-plane/src/server/services/eventBus.ts
import { EventEmitter } from "node:events";
import type { FleetEvent, FleetEventType } from "@shared/events";

export class EventBus {
  private emitter = new EventEmitter();
  constructor() { this.emitter.setMaxListeners(100); }

  on<T extends FleetEventType>(
    type: T,
    handler: (e: Extract<FleetEvent, { type: T }>) => void
  ): () => void {
    this.emitter.on(type, handler as (e: FleetEvent) => void);
    return () => this.emitter.off(type, handler as (e: FleetEvent) => void);
  }

  onAny(handler: (e: FleetEvent) => void): () => void {
    this.emitter.on("*", handler);
    return () => this.emitter.off("*", handler);
  }

  emit(e: FleetEvent): void {
    this.emitter.emit(e.type, e);
    this.emitter.emit("*", e);
  }
}

File: api/server/services/event_bus.py
from __future__ import annotations
from collections import defaultdict
from typing import Callable
from api.shared.events import FleetEvent, FleetEventType

Handler = Callable[[FleetEvent], None]


class EventBus:
    def __init__(self) -> None:
        self._typed: dict[str, list[Handler]] = defaultdict(list)
        self._any: list[Handler] = []

    def on(self, event_type: FleetEventType, handler: Handler) -> Callable[[], None]:
        self._typed[event_type].append(handler)
        def off() -> None:
            try:
                self._typed[event_type].remove(handler)
            except ValueError:
                pass
        return off

    def on_any(self, handler: Handler) -> Callable[[], None]:
        self._any.append(handler)
        def off() -> None:
            try:
                self._any.remove(handler)
            except ValueError:
                pass
        return off

    def emit(self, event: FleetEvent) -> None:
        for h in list(self._typed.get(event.type, [])):
            try:
                h(event)
            except Exception:
                pass
        for h in list(self._any):
            try:
                h(event)
            except Exception:
                pass
